api/envoy/api/v2/core/address.proto

```diff
@@ -58,9 +59,47 @@ message SocketAddress {
   bool ipv4_compat = 6;
 }
 
+// A range (inclusive) of port values
+message PortRange {
+  // First element in the port range.
+  uint32 port_value_start = 1 [(gogoproto.nullable) = false];
+  // Last element in the port range.
+  uint32 port_value_end = 2 [(gogoproto.nullable) = false];
+}
+
+// A set of socket addresses that include the same host address but may have a
+// range of ports.
+message SocketAddressPortRange {
+  Protocol protocol = 1 [(validate.rules).enum.defined_only = true];
+  // The address for this socket. :ref:`Listeners <config_listeners>` will bind
+  // to the address or outbound connections will be made. An empty address is
+  // not allowed, specify ``0.0.0.0`` or ``::`` to bind any. It's still possible to
+  // distinguish on an address via the prefix/suffix matching in
+  // FilterChainMatch after connection. For :ref:`clusters
+  // <config_cluster_manager_cluster>`, an address may be either an IP or
+  // hostname to be resolved via DNS. If it is a hostname, :ref:`resolver_name
+  // <envoy_api_field_core.SocketAddress.resolver_name>` should be set unless default
+  // (i.e. DNS) resolution is expected.
+  string address = 2 [(validate.rules).string.min_bytes = 1];
+
+  PortRange port_range = 3;
+
+  // The name of the resolver. This must have been registered with Envoy. If this is
+  // empty, a context dependent default applies. If address is a hostname this
+  // should be set for resolution other than DNS. If the address is a concrete
+  // IP address, no resolution will occur.
+  string resolver_name = 4;
+
+  // When binding to an IPv6 address above, this enables `IPv4 compatibity
+  // <https://tools.ietf.org/html/rfc3493#page-11>`_. Binding to ``::`` will
+  // allow both IPv4 and IPv6 connections, with peer IPv4 addresses mapped into
+  // IPv6 space as ``::FFFF:<IPv4-address>``.
+  bool ipv4_compat = 5;
+}
+
 message BindConfig {
   // The address to bind to when creating a socket.
-  SocketAddress source_address = 1
```

Comment:
As nice as this is, I believe it's subject to the "no breaking changes" policy

https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md ->
As of the 1.3.0 release, the Envoy user-facing configuration is locked and we will not make breaking changes between official numbered releases. This includes JSON configuration, REST/gRPC APIs (SDS, CDS, RDS, etc.), and CLI switches. We will also try to not change behavioral semantics (e.g., HTTP header processing order), though this is harder to outright guarantee.

So you could can do the uglier "extending existing config" or you can add -> deprecate -> remove across 3 envoy releases, but you can't just change things outright :-(

Reply:
Also in weighing the ugly-extend vs add->deprecate->remove, given most users won't want the port range, making it optional and something folks can ignore and put in their normal address has some value.

Reply:
Are you suggesting for "port 8000-12000" we'd do 127.0.0.1:8000 127.0.0.1:12000 and infer ranges (which doesn't seem consistent with "we can fail over to source IPs" and seems kind of messy) or "we'd have an entry for every possible address port combination for the range" which I think wouldn't scale well.

Reply:
Can we really change to repeated without changing the format? Wouldn't the json format change to [array]?

Reply:
@ggreenway you can definitely do this on the proto3 binary wire format level. For JSON/YAML, I think we'd need to teach RapidJSON to treat single values as singleton arrays if it doesn't already.

@alyssawilk Ack, for large ranges that would be hard. What about taking an `envoy.api.v2.core.Address` for the IP as we do today, and adding only the portrange?

Reply:
According to https://developers.google.com/protocol-buffers/docs/proto3#json "repeated" has a different canonical representation in JSON. I thought we had committed to not changing this? Or are you saying we should make the consumer do an implicit cast if needed?  If so, we should do that before we modify any of the .proto files in this way, right?

Reply:
I think repeated isn't an open given scaling but if it were we could require a regression test check in a hard-coded config now, pull and verify it doesn't break) for futureproofing.

yeah, having an optional range which "overrides" any address port is the "ugly extending existing config" option

Reply:
Whoops, that's a really good point about non-breaking changes.  Thinking about use cases, I'm inclined to agree that we should allow people a simple way to specify a single port that doesn't require them to engage with the range concept, which in combination moves me to the "ugly extending existing config" option--IMO, it's not worth doing the add/deprecate/remove dance for an API that is slightly simpler in itself but causes grief for people who don't want the new concept.  If anyone objects to going in this direction, let me know--I'll put a stake in the ground and say that's what I'm going to do.

Despite that I'm going to ask: Do we know how commonly used the "bind single port" feature is, and by who?  The above reasoning assumes that there's some large base of people who will want to bind to a single port who won't want to bind to a range, and I just want to make sure that's true.

@htuch: I thought about extending this concept to multiple sockets at different IP addresses, but as you and Alyssa have fleshed out, the concept of range doesn't translate well to IPs, and I'm reluctant to add API complexity without a motivating use case.